base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -667,70 +704,76 @@ impl LMDBDatabase {
-                    {
-                        return Err(ChainStorageError::InvalidOperation(
-                            "Cannot delete header because there are utxos linked to it".to_string(),
-                        ));
-                    }
+        let header =
+            self.fetch_last_header_in_txn(&txn)
+                .or_not_found("BlockHeader", "height", "last_header".to_string())?;
+        if header.height != height {
+            return Err(ChainStorageError::InvalidOperation(format!(
+                "Attempted to delete a header at height {} that was not the last header (which is at height {}). \
+                 Headers must be deleted in reverse order.",
+                height, header.height
+            )));
+        }
 
-                    lmdb_delete(&txn, &self.block_hashes_db, &hash)?;
-                    lmdb_delete(&txn, &self.headers_db, &k)?;
-                    lmdb_delete(&txn, &self.header_accumulated_data_db, &k)?;
-                    lmdb_delete(&txn, &self.kernel_mmr_size_index, &v.kernel_mmr_size.to_be_bytes())?;
-                    lmdb_delete(&txn, &self.output_mmr_size_index, &v.output_mmr_size.to_be_bytes())?;
-                }
-            },
-            DbKey::BlockHash(_) => {
-                unimplemented!("Not supported. Use delete by height");
-            },
-            DbKey::OrphanBlock(k) => {
-                if let Some(orphan) = lmdb_get::<_, Block>(&txn, &self.orphans_db, &k)? {
-                    let parent_hash = orphan.header.prev_hash;
-                    lmdb_delete_key_value(&txn, &self.orphan_parent_map_index, parent_hash.as_slice(), &k)?;
-                    let tip: Option<Vec<u8>> = lmdb_get(&txn, &self.orphan_chain_tips_db, &k)?;
-                    if tip.is_some() {
-                        if lmdb_get::<_, Block>(&txn, &self.orphans_db, parent_hash.as_slice())?.is_some() {
-                            lmdb_insert(
-                                &txn,
-                                &self.orphan_chain_tips_db,
-                                parent_hash.as_slice(),
-                                &parent_hash,
-                                "orphan_chain_tips_db",
-                            )?;
-                        }
-                        lmdb_delete(&txn, &self.orphan_chain_tips_db, &k)?;
-                    }
-                    lmdb_delete(&txn, &self.orphans_db, k.as_slice())?;
+        // TODO: This can maybe be removed for performance if the check for block_accumulated_data existing is
+        // sufficient
+
+        let hash = header.hash();
+        // Check that there are no utxos or kernels linked to this.
+
+        if !lmdb_fetch_keys_starting_with::<TransactionKernelRowData>(hash.to_hex().as_str(), &txn, &self.kernels_db)?
+            .is_empty()
+        {
+            return Err(ChainStorageError::InvalidOperation(
+                "Cannot delete header because there are kernels linked to it".to_string(),
```

Comment:
nit: more descriptive error messages

```suggestion
                format!("Cannot delete header {} ({}) because there are kernels linked to it", header.height, hash.to_hex())
```